src/dstack/_internal/server/services/logs.py

```diff
@@ -1,117 +1,342 @@
+import atexit
 import base64
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
 from datetime import datetime, timezone
 from pathlib import Path
-from typing import List
+from typing import Iterator, List, Optional, Set, TypedDict, Union
 from uuid import UUID
 
-from dstack._internal.core.models.logs import JobSubmissionLogs, LogEvent, LogEventSource
+from dstack._internal.core.models.logs import (
+    JobSubmissionLogs,
+    LogEvent,
+    LogEventSource,
+    LogProducer,
+)
 from dstack._internal.server import settings
 from dstack._internal.server.models import ProjectModel
 from dstack._internal.server.schemas.logs import PollLogsRequest
 from dstack._internal.server.schemas.runner import LogEvent as RunnerLogEvent
+from dstack._internal.utils.logging import get_logger
 
+BOTO_AVAILABLE = True
+try:
+    import boto3
+    import botocore.exceptions
+except ImportError:
+    BOTO_AVAILABLE = False
 
-def write_logs(
-    project: ProjectModel,
-    run_name: str,
-    job_submission_id: UUID,
-    runner_logs: List[RunnerLogEvent],
-    job_logs: List[RunnerLogEvent],
-):
-    if len(runner_logs) > 0:
-        runner_log_file_path = _get_runner_log_file_path(
-            project_name=project.name,
-            run_name=run_name,
-            job_submission_id=job_submission_id,
-        )
-        _write_logs(
-            log_file_path=runner_log_file_path,
-            log_events=runner_logs,
-        )
-    if len(job_logs) > 0:
-        job_log_file_path = _get_job_log_file_path(
-            project_name=project.name,
-            run_name=run_name,
-            job_submission_id=job_submission_id,
+logger = get_logger(__name__)
+
+
+def _unix_time_ms_to_datetime(unix_time_ms: int) -> datetime:
+    return datetime.fromtimestamp(unix_time_ms / 1000, tz=timezone.utc)
+
+
+def _datetime_to_unix_time_ms(dt: datetime) -> int:
+    return int(dt.timestamp() * 1000)
+
+
+def _b64encode_raw_message(message: bytes) -> str:
+    return base64.b64encode(message).decode()
+
+
+class LogStorageError(Exception):
+    pass
+
+
+class LogStorage(ABC):
+    @abstractmethod
+    def poll_logs(self, project: ProjectModel, request: PollLogsRequest) -> JobSubmissionLogs: ...
```

Comment:
Not against ellipsis per se, but maybe use `pass` for consistency?

Reply:
Ellipsis denotes `abstractmethod`s without implementation (`poll_logs` and `write_logs`), as it used in the Python documentation. `close`, on the other hand, is not abstract and has default implementation, thus `pass`.

But since only `pass` is used elsewhere in the project, I've replaced `...` with `pass` for consistency